luarules/gadgets/game_disable_unit_sharing.lua

```diff
@@ -17,21 +17,65 @@ if not gadgetHandler:IsSyncedCode() then
 	return false
 end
 
-if not (Spring.GetModOptions().disable_unit_sharing
-	-- tax force enables this
-	or (Spring.GetModOptions().tax_resource_sharing_amount or 0) ~= 0)
-	-- unit market handles the restriction instead if enabled so that selling still works
-	or Spring.GetModOptions().unit_market then
+local tax_resource_sharing_enabled = Spring.GetModOptions().tax_resource_sharing_amount ~= nil and Spring.GetModOptions().tax_resource_sharing_amount > 0
+local disable_share_econ_and_lab = Spring.GetModOptions().disable_unit_sharing_economy_and_production or tax_resource_sharing_enabled
+local disable_share_combat_units = Spring.GetModOptions().disable_unit_sharing_combat_units
+local disable_share_all = Spring.GetModOptions().disable_unit_sharing_all
+
+if not disable_share_econ_and_lab and not disable_share_combat_units and not disable_share_all then 
 	return false
 end
 
 
-function gadget:AllowUnitTransfer(unitID, unitDefID, fromTeamID, toTeamID, capture)
 
+local isEconOrLab = {} 
+local isCombatUnitOrTacticalBuilding = {} 
+
+for unitDefID, unitDef in pairs(UnitDefs) do
+	-- Mark econ units
+	if unitDef.customParams.unitgroup == "energy" or unitDef.customParams.unitgroup == "metal" then
+		isEconOrLab[unitDefID] = true
+	elseif unitDef.canResurrect then
+		isEconOrLab[unitDefID] = true
+
+
+	-- Mark labs and mobile production
+	elseif unitDef.isFactory or unitDef.isBuilder then
+		isEconOrLab[unitDefID] = true
+	end
+
+	-- Mark combat units and tactical buildings
+	if unitDef.isBuilding and not isEconOrLab[unitDefID] then 
+		isCombatUnitOrTacticalBuilding[unitDefID] = true
+	elseif #unitDef.weapons > 0 then
+		isCombatUnitOrTacticalBuilding[unitDefID] = true
+	end
+end
+
+
+
+
+
+-- Returns whether the unit is allowed to be shared according to the unit sharing restrictions.
+local function unitTypeAllowedToBeShared(unitDefID)
+	if disable_share_all then return false end
+	if disable_share_econ_and_lab and isEconOrLab[unitDefID] then return false end
+	if disable_share_combat_units and isCombatUnitOrTacticalBuilding[unitDefID] then return false end
+	return true
+end
+GG.disable_unit_sharing_unitTypeAllowedToBeShared = unitTypeAllowedToBeShared
+
+if Spring.GetModOptions().unit_market then
+	-- let unit market handle unit sharing so that buying units will still work. 
+	return false
+end
+
+
+function gadget:AllowUnitTransfer(unitID, unitDefID, fromTeamID, toTeamID, capture)
 	if(capture) then
 		return true
 	end
-	return false
+	return allowedToBeShared(unitDefID)
```

Comment:
Where does this come from???

Reply:
https://github.com/beyond-all-reason/Beyond-All-Reason/pull/4010/files/b46f7c85007f2123fe58a5b6426a9e63f14f1e43#diff-40c052408aba0b8d61ccb4d5dbd6743bc00d82a1dc532342290f5c0587c7f8c8R59-R66

Unit market uses transfer unit, which would be blocked if disable_unit_sharing always returns false in AllowUnitTransfer.

Pre-change: if unit market is enabled, don't load Disable Unit Sharing, unit market blocks all transfers except for market buy/sell
Post-change: same, but use allowedToBeShared(unitDefID) to support finer control of unit sharing.

Reply:
Unit market uses transfer unit, which would be blocked if disable_unit_sharing always returns false in AllowUnitTransfer.

Pre-change: if unit market is enabled, don't load Disable Unit Sharing, unit market blocks all transfers except for market buy/sell
Post-change: same, but use allowedToBeShared(unitDefID) to support finer control of unit sharing.